### PR TITLE
plugin: add new `plugin.query` callback

### LINKED ISSUE #263: plugin: need a callback that will return plugin's internal data
As a result of trying to debug #262, perhaps a useful utility for the multi-factor priority plugin to have would be a callback that could return all data that the plugin has at a given moment, which would include all:

- users and bank information
- users' current job limits
- users' default banks

This data should be returned in a readable format and should give us as much helpful information regarding currently running jobs and any held jobs as possible.

### PR DESCRIPTION
**Pull request by cmoussa1**
#### Background

As described in #263, it would be useful to have a utility to fetch the internal state of the multi-factor priority plugin to look at the current data it has on users, banks, and their current running and active jobs. This would be helpful when looking at jobs currently held because of a jobs limit or to see what users or banks are currently in the plugin.

---

This PR adds a new `plugin.query` callback that returns all the internal data that the plugin has about its users, banks, and their running and active job counts at the time of the query in a JSON object. It includes data such as:

- all the users/banks it has received from the flux-accounting DB via `flux account-priority-update`
- current fair-share values, `max_run_jobs` and `max_active_jobs` counts for each user/bank
- any held job IDs for each user/bank

Fixes #263

### CODE REVIEW

**Review by grondo — COMMENTED** [review]
Actually, it would be much better to just emit the JSON object on `stdout`. Then the output can easily be parsed by filters like `jq`. If a user wants to send the output to a file they can easily do so with shell redirection.

This greatly simplifies the "command" since you don't have to support any options processing. (In fact it will basically be one line)

Another note, instead of an RPC that is called `print_map_info` it might be more idiomatic to use `get_map_info` or `dump_map_info` since the RPC is not printing the map info, but just fetching it.

**Comment by cmoussa1** [comment]
Oh! Okay, that does make it easier :-) thanks for this suggestion @grondo! I'll work on changing this to just print to `stdout`.

**Comment by grondo** [comment]
@cmoussa1: @garlick had the great idea to add a new `flux jobtap info [PLUGIN]` command for jobtap and a corresponding jobtap callback from which a plugin can optionally generate detailed info to return to the caller.

Let's pause this PR until we have that support merged in flux-core. (Your work on generating the JSON object representation of the plugin internal state won't be wasted though)

**Comment by cmoussa1** [comment]
Okay! Thanks for letting me know about that. Good idea @garlick :-)

**Comment by grondo** [comment]
@cmoussa1 - check out flux-framework/flux-core#4507 for a proposed way to do this with `flux jobtap query mf_priority.so`. I'll request a review from you.

**Comment by cmoussa1** [comment]
OK, I think I've hopefully made the right adjustments to this PR after flux-framework/flux-core#4507, so I'll take this out of [WIP] and mark it as ready for some initial review whenever convenient.

File: src/plugins/mf_priority.cpp
Comment (grondo) [inline src/plugins/mf_priority.cpp:243]: I was going to suggest that return codes be checked here, but maybe this would be simpler if you created `bank_info` with `json_pack()`? e.g.:

```c
    if (!(bank_info = json_pack ("{s:s s:f s:i s:i s:i s:i}",
                                 "bank", b.first.c_str (),
                                 "fairshare", b.second.fairshare,
                                 etc....    
```

File: src/plugins/mf_priority.cpp
Comment (grondo) [inline src/plugins/mf_priority.cpp:267]: Missing `json_decref (all_users)`

**Review by grondo — COMMENTED** [review]
I've just done a quick pass and noted a couple things.

File: src/plugins/mf_priority.cpp
Comment (cmoussa1) [reply src/plugins/mf_priority.cpp]: Ah, yeah, that might be simpler. I just pushed up a `fixup` commit that just uses `json_pack ()` as well as adds a `json_decref ()` to the JSON objects that are created. It definitely does reduce the number of lines needed to create the object. Let me know if that's what you had in mind (and/or if I didn't use the `decref ()` call correctly 😬)

File: src/plugins/mf_priority.cpp
Comment (grondo) [inline src/plugins/mf_priority.cpp:243]: Did you mean `json_decref (held__jobs)` here?

**Review by grondo — CHANGES_REQUESTED** [review]
Be sure to check and handle error return from `json_object ()`, `json_array ()`, `json_object_set_new ()`, etc.

**Comment by cmoussa1** [comment]
thanks for the quick reviews @grondo! I've squashed and force-pushed your suggestions to add checks to the `json_* ()` calls in the callback, as well as changed the `json_decref ()` to `held_jobs` instead of `bank_info`. I just returned `-1` if any of the `json_* ()` calls fail.

**Review by grondo — COMMENTED** [review]
More suggestions inline :smile: Hope this helps!

**Comment by cmoussa1** [comment]
Thanks for all of the reviews and suggestions so far @grondo - can you tell I've not worked with Jannson much? 😄

I just added a `fixup` commit to try and abstract both the construction of a "user" JSON object and any subsequent "bank" JSON object(s) into their own separate functions. Hopefully this clears things up and makes it easier to do error checking.

I think the part where I find myself most confused is when to call `json_decref ()` - do you have to call that on the JSON object in the case of every error check? In the `fixup` commit I've tried to include a `json_decref ()` in every error handling body, but I'm not sure if that is necessary (or if I am `decref`-ing the right object.

**Comment by grondo** [comment]
> I think the part where I find myself most confused is when to call json_decref () - do you have to call that on the JSON object in the case of every error check? I

Yeah, the Jansson `json_t *` object reference count based memory management can get a little confusing. It may be helpful to read through the [Reference Count](https://jansson.readthedocs.io/en/2.10/apiref.html#apiref-reference-count) section of the Jansson docs first to get a basic understanding.

Basically, when you create a new `json_t` object via `json_object ()`, `json_array ()`, `json_integer ()`, and so on, libjansson is allocating some memory for you, and returning a reference to that memory as a `json_t *`. The reference count for the `json_t` starts at 1. If you don't call `json_decref ()`, or pass the reference to another object (e.g. via `json_array_append_new ()` or `json_pack()` with the little `o`), then the memory can't be freed.

Therefore as you accumulate json objects in a function with a reference count, you have to ensure you decref those objects before leaving the function due to an error. If you look at a lot of flux-core code, you will see we use a `goto error:` idiom and call `json_decref ()` on all variables that could possibly still have a reference. This sometimes allows you to reduce the code duplication.

This is also why I suggest encapsulating the creation of each inner JSON object into its own function. This makes error handling easier in that you just have to make sure the components of that one object are freed before returning an error.

You do not have to `json_decref (object)` if the object failed to get allocated, e.g. as you do currently, e.g.

```c
banks = json_array ();
if (!banks)
    json_decref (banks)
```

Since `banks == NULL`, `json_decref (banks)` does nothing and is not needed.

However, later, if `bank_info` fails to be created, you return without calling `json_decref (banks)`. Hopefully you now can see that this is a leak, since `banks` is definitely allocated at this point.

Hopefully this short description and the Jansson documentation helps. Then you can take another crack at it :-)

**Comment by cmoussa1** [comment]
OK, I've squashed and force-pushed a change to hopefully and correctly use `json_decref ()` in the functions that create the JSON objects. Like you mentioned in your suggestions above, I created a `goto error` idiom that `decref`s the `json_t` objects that are created in the two helper functions in the case when an error occurs. Let me know if this is better and/or I need to make additional changes. Thanks for helping me work through this! :P

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:228]: This looks correct to me.

I do have one style suggestion: declare `jobid` within the for loop and instead of calling `json_decref (jobid)` at the `error:` label,  use
```c
if (json_array_append_new (held_jobs, jobid) < 0) :
    json_decref (jobid)
    goto error;
}
```
(and skip the `json_decref (jobid)` at the bottom of the for loop).

This is a simple case but I find it gets really confusing in general when you allocate something in a loop and free it outside of the loop.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:244]: Need to handle the case where `add_held_jobs()` returns NULL.

**Review by garlick — COMMENTED** [review]
Just revewing that last fixup commit since it was the result of some offline discussion.  I think you've nearly got it @cmoussa1.

File: src/plugins/mf_priority.cpp
Comment (cmoussa1) [reply src/plugins/mf_priority.cpp]: Ah, I like this style suggestion - thanks. This seems more straightforward to me to declare and free the variable within the for-loop, so I'll adopt this!

File: src/plugins/mf_priority.cpp
Comment (cmoussa1) [reply src/plugins/mf_priority.cpp:244]: Yes, thanks for pointing this out - I just squashed and pushed a fix to add a check for `held_jobs` after this function call to check for `NULL`.

**Comment by cmoussa1** [comment]
Thanks for the great suggestions @garlick - I just squashed and pushed based on the feedback you gave above. Let me know what you think!

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:222]: Use `json_array_append_new()` here.  The distinction is that `_new()` doesn't take a reference on `jobid`.  It "steals" the existing reference, so once that call is successful, you no longer have to worry about freeing `jobid`.  Conversely, when it fails you do.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:286]: Need to check for NULL failure here.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:284]: Like the other function, you _could_ declare `bank_info` inside the loop, then use `json_array_append_new()` as described earlier.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:323]: Need to check for NULL return from `banks_to_json()`.

Also if the `json_object_set_new()` on "banks" fails, now userid is owned by user, so you will get a double free on the error path (userid is freed once when user is freed, then again explicitly).

**Review by garlick — COMMENTED** [review]
Oops, some of the same comments I made earlier apply to the other functions.  Sorry I made the one comment standalone - I meant to group all these together in one review.

File: src/plugins/mf_priority.cpp
Comment (cmoussa1) [reply src/plugins/mf_priority.cpp]: I see. Thanks for pointing this out! There are two other calls to `json_array_append ()` in the helper functions (in `banks_to_json ()` and `query_cb ()`, but I left those as is because I'm pretty sure I free the JSON object after the call to `array_append ()`. Let me know if those would be better to also change to `array_append_new ()`.

File: src/plugins/mf_priority.cpp
Comment (cmoussa1) [reply src/plugins/mf_priority.cpp:323]: Ah, okay. Maybe I should just declare a separate JSON object to put the result of `banks_to_json ()` into. Then I can check for `NULL` before adding it with `json_object_set_new ()`.

> Also if the `json_object_set_new()` on "banks" fails, now userid is owned by user, so you will get a double free on the error path (userid is freed once when user is freed, then again explicitly).

I think I see what you're saying. So in this case, I do not need to explicitly free `userid` in the error path since it is already owned by `user`?

```diff
error:
    json_decref (user);
-   json_decref (userid);
    return NULL;
```

Comment (garlick) [reply src/plugins/mf_priority.cpp:323]: Yes but you will need to free it if `json_object_set_new()` on `userid` fails.

**Comment by cmoussa1** [comment]
Thanks for taking another look at this and providing some more feedback - I just squashed and force-pushed based on the above suggestions. 

I think I should also change the following based on some of the feedback above:

```c++
    for (auto const &u : users) {
        if (json_array_append (all_users, user_to_json (p, u)) < 0) {
            json_decref (all_users);
            return -1;
        }
    }
```

`user_to_json (p, u)` should also be checked for NULL before appending it to the `all_users` JSON_array.

**Comment by cmoussa1** [comment]
Just squashed and pushed the change to check the return value for `user_to_json (p, u)` and to change to use `json_array_append_new ()`.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:242]: Oops I think this was left over from earlier?   The `bank_info` variable is overwritten by `json_pack()` further down, so the reference to the object created here is leaked.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:263]: bank_info will always be NULL here once the extraneous `json_object()` call is removed,  so this line can be removed.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:285]: The reference to the object created here is overwritten by `pack_bank_info_object()` and leaked.  This should be dropped.

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:320]: This error path leaks `user`.  Use `goto error` here.

**Review by garlick — COMMENTED** [review]
Getting very close, but there remain a few things to fix.

**Comment by cmoussa1** [comment]
Thanks @garlick! I just squashed and force pushed based on your suggestions above

File: src/plugins/mf_priority.cpp
Comment (garlick) [inline src/plugins/mf_priority.cpp:357]: Need to `json_decref (all_users)` before returning -1 here, or it leaks.

**Review by garlick — APPROVED** [review]
Just one more memory leak and LGTM!  I'll approve and you can set MWP once you've addressed that.

**Comment by grondo** [comment]
Probably should change the PR title prefix from `cmd:` to `plugin:`?

**Comment by cmoussa1** [comment]
> Probably should change the PR title prefix from `cmd:` to `plugin:`?

Totally, thanks for pointing this out. I forgot to change it after flux-framework/flux-core#4507.

> Just one more memory leak and LGTM! I'll approve and you can set MWP once you've addressed that.

Thanks for taking the time to help review this PR! Really appreciate all the feedback and suggestions - I just pushed up a fix to address the above suggestion, and will set MWP shortly.

**Comment by cmoussa1** [comment]
Ah, perhaps this hasn't landed yet because of your requested changes @grondo? Is it OK with you if I remove you as a reviewer? Or you could take a look and give your approval too. :-)

**Review by grondo — APPROVED** [review]
Just tested this out and LGTM! Nice work! Ran the test with `FLUX_TEST_VALGRIND=t` and no leaks or errors detected.

**Comment by grondo** [comment]
Weird, some old comments popped up when I added that approval comment. Those were stale from a previous review so can be ignored.

**Pull request merged**
